test/generator/stubtest_jittest.cpp

```diff
@@ -60,56 +59,136 @@ int main(int argc, char **argv) {
-    verify(buffer_input, 1.f, 0, b1);
-
-    Realization static_compiled_buffer_output_realized = gen.static_compiled_buffer_output.realize({kSize, kSize, 3});
-    Buffer<uint8_t, 3> b2 = static_compiled_buffer_output_realized;
-    verify(buffer_input, 1.f, 42, b2);
-
-    for (int i = 0; i < 2; ++i) {
-        Realization array_buffer_output_realized = gen.array_buffer_output[i].realize({kSize, kSize, 3});
-        Buffer<uint8_t, 3> b2 = array_buffer_output_realized;
-        verify(buffer_input, 1.f, 1 + i, b2);
+    // Alternately, instead of using Generator Stubs, we can just use the Callable interface.
+    // We can call this on any Generator that is registered in the current process.
+    {
+        Buffer<uint8_t, 3> buffer_input = make_image<uint8_t>(0);
+        Buffer<float, 3> simple_input = make_image<float>(0);
+        Buffer<float, 3> array_input0 = make_image<float>(0);
+        Buffer<float, 3> array_input1 = make_image<float>(1);
+        Buffer<float, 3> typed_buffer_output(kSize, kSize, 3);
+        Buffer<float, 3> untyped_buffer_output(kSize, kSize, 3);
+        Buffer<float, 3> tupled_output0(kSize, kSize, 3);
+        Buffer<int32_t, 3> tupled_output1(kSize, kSize, 3);
+        Buffer<uint8_t, 3> array_buffer_input0 = make_image<uint8_t>(0);
+        Buffer<uint8_t, 3> array_buffer_input1 = make_image<uint8_t>(1);
+        Buffer<float, 3> simple_output(kSize, kSize, 3);
+        // TODO: see Issues #3709, #3967
+        Buffer<void, 3> float16_output(halide_type_t(halide_type_float, 16), kSize, kSize, 3);
+        Buffer<void, 3> bfloat16_output(halide_type_t(halide_type_bfloat, 16), kSize, kSize, 3);
+        Buffer<float, 3> tuple_output0(kSize, kSize, 3), tuple_output1(kSize, kSize, 3);
+        Buffer<int16_t, 3> array_output0(kSize, kSize, 3), array_output1(kSize, kSize, 3);
+        Buffer<uint8_t, 3> static_compiled_buffer_output(kSize, kSize, 3);
+        Buffer<uint8_t, 3> array_buffer_output0(kSize, kSize, 3), array_buffer_output1(kSize, kSize, 3);
+
+        // Note that this Generator has several GeneratorParams that need to be set correctly
+        // before compilation -- in the Stub case above, the values end up being inferred
+        // from the specific inputs we provide, but for the JIT (and AOT) cases, there are
+        // no such inputs available, so we must be explicit. (Note that these are the same
+        // values specified in our Make/CMake files.)
+        const std::map<std::string, std::string> gp = {
+            {"untyped_buffer_input.type", "uint8"},
+            {"untyped_buffer_input.dim", "3"},
+            {"simple_input.type", "float32"},
+            {"array_input.type", "float32"},
+            {"array_input.size", "2"},
+            {"int_arg.size", "2"},
+            {"tuple_output.type", "float32,float32"},
+            {"vectorize", "true"},
+        };
+
+        Callable stubtest = create_callable_from_generator(context, "stubtest", gp);
```

Comment:
Referring to the registered generator by string name seems a bit weird here. Given a Generator class (not a stub), is it possible to say something akin to MyGeneratorClass::create(target).compile_to_callable()? Looking through Generator.h it seems like the awkward bit is setting the generator params

Reply:
You have a point -- I did it this way because for C++ Generators, there usually is no in-scope reference to the Generator class itself (it's often in a .cpp file with no .h, in an anon namespace to boot), so doing a lookup by name is simpler, and since it's still resolved at Halide Compilation time, an assertion-failure from name lookup is tantamount to any other compilation error.

That said, it would be easy enough to add such a method so that C++ Generators that *are* declared in-scope could be used directly. 

Also, maybe it would be feasible to just allow an `extern` declaration of the C++ Generator (or its factory or whatnot) to avoid needing .h files and just relying on name-mangling to keep the build honest -- though extern declaration of a CRTP template is kinda ugly. (Yeah, any sort of extern-declaration in a .cpp file is pretty gross, but this is C++ we're talking about, it's not like we have non-gross options for this sort of thing.)

Reply:
> maybe it would be feasible to just allow an extern declaration of the C++ Generator

So the answer here is "No, not really": forward declaration of a template class defined via CRTP is, well, honestly, I haven't figured out how to do that. But I'm not sure that would help much, since we'd still just have a class forward declaration without any "guts" to it (ie no methods to call). 

So I guess we're really looking at three possibilities here:
- Your Generator is defined in a such a way that its class definition is fully visible at the call site (either it's in a .h file or you have multiple Generators in a single .cpp). Using your approach here would indeed work and arguably be simplest, but AFAIK literally zero of the Generators inside Google fit this pattern, and only a tiny handful in existing public Halide code. If we want to encourage people to put Generators into .h files (instead of .cpp files) then this makes sense for future code (and we have no good reason not to support it in addition to other stuff), but as of right now it would seem to be of extremely limited use.
- The existing looking-by-string approach I have in place now. This is unsatisfying for the reasons you list, but works (well) in all the existing use cases I know of.
- The halfway approach... this is kinda yucky but here goes:
    - We rely on the fact that `HALIDE_REGISTER_GENERATOR()` is a macro and can define hidden helpers (and that this macro is *always* in the global namespace
    - We add a macro like `HALIDE_EXTERN_GENERATOR_FACTORY()` which is just the extern-declaration for those hidden helper funcs
    - Then we can use it something like
    ```
    // file: example_generator.cpp
    class ExampleClassName : Generator<ExampleClassName> {...};
    HALIDE_REGISTER_GENERATOR(ExampleClassName, example_registered_name)

    // example_jittest.cpp
    // (Note that we have to use the 'registration' name here, not the class name,
    // since the class name may well be in an anonymous namespace that is literally
    // impossible to specify here.)
    auto factory = HALIDE_EXTERN_GENERATOR_FACTORY(example_registered_name);

    auto callable = factory(context)->compile_to_callable();

    ```
    Stylistically, this isn't really any better than the current "look up by string" approach in this PR, but it would have the effect of resolving issues at link time, rather than compilation time. What do you think?